File: Java/src/main/java/QuestionBank/Solution701_800/Solution733.java
package QuestionBank.Solution701_800;

import java.util.Arrays;

/**
 * @author windows
 */
public class Solution733 {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        boolean[][] used = new boolean[image.length][image[0].length];
        floodFill(image, sr, sc, color, image[sr][sc], used);
        return image;
    }

    private void floodFill(int[][] image, int sr, int sc, int color, int target, boolean[][] used) {
        if (sr >= 0 && sr < image.length && sc >= 0 && sc < image[0].length) {
            if (image[sr][sc] == target && used[sr][sc] == false) {
                image[sr][sc] = color;
                used[sr][sc] = true;
                floodFill(image, sr, sc + 1, color, target, used);
                floodFill(image, sr, sc - 1, color, target, used);
                floodFill(image, sr + 1, sc, color, target, used);
                floodFill(image, sr - 1, sc, color, target, used);
            }
        }
    }


    public static void main(String[] args) {
        Solution733 solution733 = new Solution733();
        int[][] image = new int[][]{{0, 0, 0}, {1, 0, 0}};
        System.out.println(Arrays.deepToString(solution733.floodFill(image, 1, 0, 2)));
    }
}

File: C++/src/Solution701_800/Solution733.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution733.h"

vector<vector<int>> Solution733::floodFill(vector<vector<int>> &image, int sr, int sc, int color) {
    int currColor = image[sr][sc];
    if (currColor != color) {
        dfs(image, sr, sc, currColor, color);
    }
    return image;
}

void Solution733::dfs(vector<vector<int>> &image, int x, int y, int currColor, int color) {
    if (image[x][y] == currColor) {
        image[x][y] = color;
        for (int i = 0; i < 4; i++) {
            int mx = x + dx[i], my = y + dy[i];
            if (mx >= 0 && mx < image.size() && my >= 0 && my < image[0].size()) {
                dfs(image, mx, my, currColor, color);
            }
        }
    }
}

